tools/config/__init__.py

```diff
@@ -710,8 +711,23 @@ def _generate_bootloader_build(self, rom_start, rom_size):
             if part.minaddr() != rom_start:
                 raise ConfigException("bootloader executable does not "
                                       "start at 0x%x" % rom_start)
-            part_size = (part.maxaddr() - part.minaddr()) + 1
-            part_size = Config._align_ceiling(rom_start + part_size, self.sectors) - rom_start
+            if len(part.segments()) == 1:
+                part_size = (part.maxaddr() - part.minaddr()) + 1
+                part_size = Config._align_ceiling(rom_start + part_size, self.sectors) - rom_start
+            else:
+                # assume first segment is at start
+                # second at the end or outside ROM
+                # rest outside regular ROM
+                if part.segments()[0][0] != 0x0:
+                    raise ConfigException("bootloader segments not in order")
+                part_size = part.segments()[0][1] - part.segments()[0][0]
+                part_size = Config._align_ceiling(rom_start + part_size, self.sectors) - rom_start
+                if part.segments()[1][0] < rom_end and self.target.restrict_size is None:
+                    if self.target.app_offset:
```

Comment:
`target.app_offset` is not the only way that an application can be pushed further back: `target.header_format` and `target.header_offset` can also push an application further back. Either this check should check for all of them, or `max_app_addr` should be used unconditionally.